File: Graph/Graph_01.java
/*        (5)
      0--------1
        (1) /   \   (3)
           /     \
          2-------3
           \  (1)
     (2)    \
             4
 */

 // This code defines a simple graph structure using an adjacency list representation.
 // we are using one array of ArrayList which contains an edge . each edge has 3 properties , one src , dist and weight

import java.util.*;
import java.io.*;
import java.math.*;

 public class Graph {
    static class Edge {
        int src;
        int dest;
        int weight;

        // Constructor to initialize the edge
        // src: source vertex, dest: destination vertex, weight: weight of the edge
        // This constructor initializes the edge with the given source, destination, and weight.
        // It is used to create an edge in the graph with the specified properties.
        // The src and dest represent the vertices connected by the edge, and weight represents the cost or distance associated with that edge.
        
        public Edge(int src, int dest, int weight) {
            this.src = src;
            this.dest = dest;
            this.weight = weight;
            }   
        }
        public static void createGraph(ArrayList<Edge>[] graph) {
        // Initialize the adjacency list for each vertex
            for (int i = 0; i < graph.length; i++) {
                graph[i] = new ArrayList<>();
            }
            // Add edges to the graph 
            // 0-vertex
            graph[0].add(new Edge(0, 1, 1));
            graph[0].add(new Edge(0, 2, 1));
            // 1-vertex
            graph[1].add(new Edge(1, 0, 1));
            graph[1].add(new Edge(1, 3, 1));
            // 2-vertex
            graph[2].add(new Edge(2, 0, 1));
            graph[2].add(new Edge(2, 4, 1));
            // 3-vertex
            graph[3].add(new Edge(3, 1, 1));
            graph[3].add(new Edge(3, 5, 1));
            graph[3].add(new Edge(3, 4, 1));
            // 4-vertex
            graph[4].add(new Edge(4, 3, 1));
            graph[4].add(new Edge(4, 5, 1));
            // 5-vertex
            graph[5].add(new Edge(5, 3, 1));
            graph[5].add(new Edge(5, 6, 1));
            graph[5].add(new Edge(5, 4, 1));

            }



            //BFS Traversal
            public static void BFS_Traversal(ArrayList<Edge>[] graph) {     
                Queue<Integer> queue = new LinkedList<>();
                boolean[] visited = new boolean[graph.length];
                queue.add(0); // Start BFS from vertex 0

                while(!queue.isEmpty()){
                    int current = queue.remove();
                    if(!visited[current]) {
                        visited[current] = true; // Mark the current vertex as visited
                        System.out.print(current + " "); // Print the current vertex

                        // Traverse all edges from the current vertex
                        for(Edge edge : graph[current]) {
                            if(!visited[edge.dest]) {
                                queue.add(edge.dest); // Add unvisited destination to the queue
                            }
                        }
                    }
                }
            } 




            public static void DFS_Traversal(ArrayList<Edge>[] graph) { 
                   
            }
    public static void main(String[] args) {
       
   /*      
            1-------3 -------5------6
          /         \        /
        /            \      /
       0             \     /     
        \             \   /
         \             \ /   
          2-------------4

    */
             int V = 7; // Number of vertices
             ArrayList<Edge> [] graph = new ArrayList[V];
             createGraph(graph);

            System.out.println("BFS Traversal of the graph:");
            BFS_Traversal(graph);
            System.out.println();

    // for(int j=0;j<graph.length;j++){
    //     for(int i=0;i<graph[j].size();i++) {
    //         Edge e = graph[j].get(i);
    //         System.out.println("Edge from " + e.src + " to " + e.dest + " with weight " + e.weight);
    //         }
    //         System.out.println("-----------------");
    //     }
    }
}






import java.util.ArrayList;

// Edge class with source, destination, and weight
class Edge {
    int src, dest, wt;

    public Edge(int src, int dest, int wt) {
        this.src = src;
        this.dest = dest;
        this.wt = wt;
    }
}

public class Graph {
    // Function to create the graph
    public static void createGraph(ArrayList<Edge>[] graph) {
        // Initialize the array of ArrayLists
        for (int i = 0; i < graph.length; i++) {
            graph[i] = new ArrayList<>();
        }

        // Original edges
        graph[0].add(new Edge(0, 1, 5));
        graph[0].add(new Edge(0, 2, 3));
        graph[1].add(new Edge(1, 3, 6));
        graph[2].add(new Edge(2, 3, 7));
        graph[3].add(new Edge(3, 4, 4));
        graph[4].add(new Edge(4, 0, 2));

        // Additional edges
        graph[1].add(new Edge(1, 2, 1));
        graph[2].add(new Edge(2, 4, 8));
        graph[4].add(new Edge(4, 1, 9));
    }

    // Function to print the graph
    public static void printGraph(ArrayList<Edge>[] graph) {
        for (int i = 0; i < graph.length; i++) {
            for (Edge e : graph[i]) {
                System.out.println("From " + e.src + " to " + e.dest + " with weight " + e.wt);
            }
        }
    }

    public static void main(String[] args) {
        int V = 5; // Number of vertices
        ArrayList<Edge>[] graph = new ArrayList[V];

        createGraph(graph);
        printGraph(graph);
    }
}

/*
    GRAPH STRUCTURE (Directed & Weighted):

            (0)
           /   \
         5/     \3
         /       \
       (1) -----> (2)
        | 1        | \
       6|         7|  \8
        |          |   \
       (3) ------> (4)
             4      ↑
                   9|
                   (1)

    Nodes: 0, 1, 2, 3, 4
    Directed Edges with Weights:
    0 -> 1 (5), 0 -> 2 (3)
    1 -> 3 (6), 1 -> 2 (1)
    2 -> 3 (7), 2 -> 4 (8)
    3 -> 4 (4)
    4 -> 0 (2), 4 -> 1 (9)
*/







--------------------------------------------------------------





      import java.util.*;

public class GraphDFS {
    static class Graph {
        int V;
        ArrayList<ArrayList<Integer>> adj;

        // Constructor
        Graph(int V) {
            this.V = V;
            adj = new ArrayList<>();
            for (int i = 0; i < V; i++)
                adj.add(new ArrayList<>());
        }

        // Add edge
        void addEdge(int u, int v) {
            adj.get(u).add(v);
            adj.get(v).add(u); // Omit this if it's a directed graph
        }

        // DFS Utility
        void dfsUtil(int v, boolean[] visited) {
            visited[v] = true;
            System.out.print(v + " ");

            for (int neighbor : adj.get(v)) {
                if (!visited[neighbor])
                    dfsUtil(neighbor, visited);
            }
        }

        // DFS from a given start node
        void dfs(int start) {
            boolean[] visited = new boolean[V];
            System.out.print("DFS traversal starting from node " + start + ": ");
            dfsUtil(start, visited);
            System.out.println();
        }
    }

    // Main
    public static void main(String[] args) {
        Graph g = new Graph(5);
        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(1, 3);
        g.addEdge(1, 4);

        /*
         Graph visualization:

             0
            / \
           1   2
          / \
         3   4

        Output of DFS from node 0: 0 1 3 4 2
        */

        g.dfs(0);
    }
}

File: Graph/Graph_01.cpp
// /* Day-01 of learning Graphs in C++ by taught Apna-College (Shradha Khapra).
//    Graph is a network of Nodes 
// */

// // Implementation of Graph using Adjacency List

// // #include<iostream>
// // #include<vector>
// // #include<list>
// // using namespace std;
// // // creating Graph class from scratch
// // class Graph {
// //     int V; // Number of varices of the graph
// //     list<int>* l; // dynamic list (doubly-linkedlist) to store the neighbors of the graph of size V
// // public:
// // // constructor
// //     Graph(int V) {
// //         this->V = V; // putting the value of V into the parameterized constructor Graph 
// //         l = new list<int> [V]; // creating list of size V
// //     }
    
// //     // function/method to add edges 
// //     void addEdge(int u, int v){
// //         l[u].push_back(v);
// //         l[v].push_back(u);
// //     }

// //     // function to pring graph
// //     void print(){
// //         for(int u=0;u<V;u++){
// //             list<int> neighbors = l[u];
// //             cout<<u<<" : ";
// //             for(int v: neighbors){
// //                 cout<<v<<" ";
// //             }
// //             cout<<"\n";
// //         }
// //     }
// // };



// // int main(){
// //     Graph graph(5); // creating graph 
// //     graph.addEdge(0,1);// adding edges
// //     graph.addEdge(1,2);// adding edges
// //     graph.addEdge(1,3);// adding edges
// //     graph.addEdge(2,3);// adding edges
// //     graph.addEdge(2,4);// adding edges
    
// //     graph.print();// calling print function
// //     return 0;
// // };

// /* 
// output = 0 : 1
//          1 : 0 2 3
//          2 : 1 3 4
//          3 : 1 2
//          4 : 2  
// */


// // Breath-First-Search (BFS) in Graph
// // #include<iostream>
// // #include<vector>
// // #include<list>
// // #include<queue>
// // using namespace std;
// // // creating Graph class from scratch
// // class Graph {
// //     int V; // Number of varices of the graph
// //     list<int>* l; // dynamic list (doubly-linkedlist) to store the neighbors of the graph of size V
// // public:
// // // constructor
// //     Graph(int V) {
// //         this->V = V; // putting the value of V into the parameterized constructor Graph 
// //         l = new list<int> [V]; // creating list of size V
// //     }
    
// //     // function/method to add edges 
// //     void addEdge(int u, int v){
// //         l[u].push_back(v);
// //         l[v].push_back(u);
// //     }

// //     // function to pring graph
// //     void print(){
// //         for(int u=0;u<V;u++){
// //             list<int> neighbors = l[u];
// //             cout<<u<<" : ";
// //             for(int v: neighbors){
// //                 cout<<v<<" ";
// //             }
// //             cout<<"\n";
// //         }
// //     }

// //     void BFS(){
// //         queue<int> q;
// //         vector<bool> vis(V,false);
// //         q.push(0);
// //         vis[0]=true;

// //         while(q.size()>0){
// //             int u = q.front();
// //             q.pop();
// //             cout<<u<<" ";
// //             list<int> neighbors = l[u];
// //             for(int v:neighbors){
// //                 if(!vis[v]){
// //                     vis[v]= true;
// //                     q.push(v);
// //                 }
            
// //             }
// //         }
// //         cout<<"\n";
// //     }
// // };



// // int main(){
// //     Graph graph(7);      // creating graph 
// //     graph.addEdge(0,1);  // adding edges
// //     graph.addEdge(0,2);  // adding edges
// //     graph.addEdge(1,3);  // adding edges
// //     graph.addEdge(2,4);  // adding edges
// //     graph.addEdge(3,5);  // adding edges
// //     graph.addEdge(4,5);  // adding edges
// //     graph.addEdge(5,6);  // adding edges
    
// //     //graph.print(); calling print function
// //     graph.BFS();
// //     return 0;
// // };

// // /*
// //     output = 0 1 2 3 4 5 6
// // */







// // creating va graph using adjacency list in C++
// #include<iostream>
// #include<list>
// using namespace std;

// class Graph {
//     int v;
//     list<int>* l;
//     public:
//         Graph(int v) {
//         this->v=v;
//         this->l= new list<int>[v];
//         }


//         void addEdge(int u,int v){
//             l[u].push_back(v);
//             l[v].push_back(u);
//         }

//         void print() {
//             for(int i=0;i<v;i++){
//                 cout<<i<<" : ";
//                 for(int j:l[i]){
//                     cout<<j<<" ";
//                 }
//                 cout<<"\n";
//             }
//         }

//         ~Graph() {
//             delete[] l; // Free dynamically allocated memory for the adjacency list
//         }

//         // Destructor to clean up the dynamically allocated memory
//         // Note: The destructor is important to avoid memory leaks, especially in larger applications.a1
//         // If you have other resources to clean up, you can add them here as well.a1

        
        

        
        


//     };


// int main(){
//     Graph graph(5);
//     graph.addEdge(0,1);
//     graph.addEdge(1,2);
//     graph.addEdge(1,3); 
//     graph.addEdge(2,3);
//     graph.addEdge(2,4);
    
//     graph.print();
//     // Output:
//     // 0 : 1
//     // 1 : 0 2 3
//     // 2 : 1 3 4
//     // 3 : 1 2
//     // 4 : 2


//     return 0;
// }









// /* Day-01 of learning Graphs in C++ by taught Apna-College (Shradha Khapra).
//    Graph is a network of Nodes 
// */

// // Implementation of Graph using Adjacency List

// // #include<iostream>
// // #include<vector>
// // #include<list>
// // using namespace std;
// // // creating Graph class from scratch
// // class Graph {
// //     int V; // Number of varices of the graph
// //     list<int>* l; // dynamic list (doubly-linkedlist) to store the neighbors of the graph of size V
// // public:
// // // constructor
// //     Graph(int V) {
// //         this->V = V; // putting the value of V into the parameterized constructor Graph 
// //         l = new list<int> [V]; // creating list of size V
// //     }
    
// //     // function/method to add edges 
// //     void addEdge(int u, int v){
// //         l[u].push_back(v);
// //         l[v].push_back(u);
// //     }

// //     // function to pring graph
// //     void print(){
// //         for(int u=0;u<V;u++){
// //             list<int> neighbors = l[u];
// //             cout<<u<<" : ";
// //             for(int v: neighbors){
// //                 cout<<v<<" ";
// //             }
// //             cout<<"\n";
// //         }
// //     }
// // };



// // int main(){
// //     Graph graph(5); // creating graph 
// //     graph.addEdge(0,1);// adding edges
// //     graph.addEdge(1,2);// adding edges
// //     graph.addEdge(1,3);// adding edges
// //     graph.addEdge(2,3);// adding edges
// //     graph.addEdge(2,4);// adding edges
    
// //     graph.print();// calling print function
// //     return 0;
// // };

// /* 
// output = 0 : 1
//          1 : 0 2 3
//          2 : 1 3 4
//          3 : 1 2
//          4 : 2  
// */


// // Breath-First-Search (BFS) in Graph
// // #include<iostream>
// // #include<vector>
// // #include<list>
// // #include<queue>
// // using namespace std;
// // // creating Graph class from scratch
// // class Graph {
// //     int V; // Number of varices of the graph
// //     list<int>* l; // dynamic list (doubly-linkedlist) to store the neighbors of the graph of size V
// // public:
// // // constructor
// //     Graph(int V) {
// //         this->V = V; // putting the value of V into the parameterized constructor Graph 
// //         l = new list<int> [V]; // creating list of size V
// //     }
    
// //     // function/method to add edges 
// //     void addEdge(int u, int v){
// //         l[u].push_back(v);
// //         l[v].push_back(u);
// //     }

// //     // function to pring graph
// //     void print(){
// //         for(int u=0;u<V;u++){
// //             list<int> neighbors = l[u];
// //             cout<<u<<" : ";
// //             for(int v: neighbors){
// //                 cout<<v<<" ";
// //             }
// //             cout<<"\n";
// //         }
// //     }

// //     void BFS(){
// //         queue<int> q;
// //         vector<bool> vis(V,false);
// //         q.push(0);
// //         vis[0]=true;

// //         while(q.size()>0){
// //             int u = q.front();
// //             q.pop();
// //             cout<<u<<" ";
// //             list<int> neighbors = l[u];
// //             for(int v:neighbors){
// //                 if(!vis[v]){
// //                     vis[v]= true;
// //                     q.push(v);
// //                 }
            
// //             }
// //         }
// //         cout<<"\n";
// //     }
// // };



// // int main(){
// //     Graph graph(7);      // creating graph 
// //     graph.addEdge(0,1);  // adding edges
// //     graph.addEdge(0,2);  // adding edges
// //     graph.addEdge(1,3);  // adding edges
// //     graph.addEdge(2,4);  // adding edges
// //     graph.addEdge(3,5);  // adding edges
// //     graph.addEdge(4,5);  // adding edges
// //     graph.addEdge(5,6);  // adding edges
    
// //     //graph.print(); calling print function
// //     graph.BFS();
// //     return 0;
// // };

// // /*
// //     output = 0 1 2 3 4 5 6
// // */







// // creating va graph using adjacency list in C++
// #include<iostream>
// #include<list>
// using namespace std;

// class Graph {
//     int v;
//     list<int>* l;
//     public:
//         Graph(int v) {
//         this->v=v;
//         this->l= new list<int>[v];
//         }


//         void addEdge(int u,int v){
//             l[u].push_back(v);
//             l[v].push_back(u);
//         }

//         void print() {
//             for(int i=0;i<v;i++){
//                 cout<<i<<" : ";
//                 for(int j:l[i]){
//                     cout<<j<<" ";
//                 }
//                 cout<<"\n";
//             }
//         }

//         ~Graph() {
//             delete[] l; // Free dynamically allocated memory for the adjacency list
//         }

//         // Destructor to clean up the dynamically allocated memory
//         // Note: The destructor is important to avoid memory leaks, especially in larger applications.a1
//         // If you have other resources to clean up, you can add them here as well.a1

        
        

        
        








#include <iostream>
#include <vector>
#include <queue>
using namespace std;

/*
    GRAPH STRUCTURE (Undirected, Unweighted):

            0
           / \
          1   2
           \   \
            3---4
             \ /
              5
              |
              6

    Edges:
    0 - 1, 0 - 2
    1 - 3
    2 - 4
    3 - 4, 3 - 5
    4 - 5
    5 - 6

    BFS Traversal from node 0: 0 1 2 3 4 5 6  
    DFS Traversal from node 0: 0 1 3 5 6 4 2
*/

// Edge class with source, destination, and weight
class Edge {
public:
    int src, dest, weight;

    // Constructor using 'this' keyword
    Edge(int src, int dest, int weight) {
        this->src = src;
        this->dest = dest;
        this->weight = weight;
    }
};

// Create the graph as an adjacency list
void createGraph(vector<Edge> graph[], int V) {
    for (int i = 0; i < V; i++) {
        graph[i] = vector<Edge>();
    }

    // Add bidirectional edges
    graph[0].push_back(Edge(0, 1, 1));
    graph[0].push_back(Edge(0, 2, 1));

    graph[1].push_back(Edge(1, 0, 1));
    graph[1].push_back(Edge(1, 3, 1));

    graph[2].push_back(Edge(2, 0, 1));
    graph[2].push_back(Edge(2, 4, 1));

    graph[3].push_back(Edge(3, 1, 1));
    graph[3].push_back(Edge(3, 5, 1));
    graph[3].push_back(Edge(3, 4, 1));

    graph[4].push_back(Edge(4, 3, 1));
    graph[4].push_back(Edge(4, 5, 1));

    graph[5].push_back(Edge(5, 3, 1));
    graph[5].push_back(Edge(5, 6, 1));
    graph[5].push_back(Edge(5, 4, 1));
}

// BFS Traversal
void BFS_Traversal(vector<Edge> graph[], int V) {
    vector<bool> visited(V, false);
    queue<int> q;

    q.push(0); // Start from vertex 0

    while (!q.empty()) {
        int curr = q.front();
        q.pop();

        if (!visited[curr]) {
            visited[curr] = true;
            cout << curr << " ";

            for (Edge e : graph[curr]) {
                if (!visited[e.dest]) {
                    q.push(e.dest);
                }
            }
        }
    }
    cout << endl;
}

// DFS Helper (Recursive)
void DFS_Helper(int curr, vector<Edge> graph[], vector<bool>& visited) {
    visited[curr] = true;
    cout << curr << " ";

    for (Edge e : graph[curr]) {
        if (!visited[e.dest]) {
            DFS_Helper(e.dest, graph, visited);
        }
    }
}

// DFS Traversal
void DFS_Traversal(vector<Edge> graph[], int V) {
    vector<bool> visited(V, false);
    DFS_Helper(0, graph, visited);
    cout << endl;
}

// Main Function
int main() {
    int V = 7;
    vector<Edge> graph[V];

    createGraph(graph, V);

    cout << "BFS Traversal of the graph: ";
    BFS_Traversal(graph, V);

    cout << "DFS Traversal of the graph: ";
    DFS_Traversal(graph, V);

    return 0;
}
